### PR TITLE
Fixes #6 Introduce comparison tolerance

### LINKED ISSUE #6: Introduce comparison tolerance
tbd.

### PR DESCRIPTION
**Pull request by Yuri05**
Fixes #6 Introduce comparison tolerance

### CODE REVIEW

**Comment by Yuri05** [comment]
Currently only tests for ODE variables are implemented.
Should also add some tests for observers

File: src/OSPSuite.SimModel/src/Species.cpp [hunk 1]
```diff
@@ -267,6 +267,9 @@ void Species::RescaleValues ()
 	
 	for(int i =0; i<_valuesSize;i++)
 		_values[i] *= m_ODEScaleFactor;
+
+	//threshold for numeric comparisons must be rescaled the same way!
+	_comparisonThreshold *= m_ODEScaleFactor;
```
Comment (msevestre) [inline src/OSPSuite.SimModel/src/Species.cpp]: Don't understand that. Also the default value is 0 . Are we sure that SetComparisonThreshold is called BEFORE RescaleValues?

File: tests/OSPSuite.SimModel.Tests/src/SimulationSpecs.cpp [hunk 2]
```diff
@@ -648,6 +663,22 @@ namespace SimulationTests
...
+			const double AbsTol = sut->UsedAbsoluteTolerance;
+
+			//expected threshold for ode variables
+			const double threshold = 10.0 * AbsTol;
+
+			BDDExtensions::ShouldBeEqualTo(_y1->GetComparisonThreshold(), threshold * _scaleFactor);
+			BDDExtensions::ShouldBeEqualTo(_y2->GetComparisonThreshold(), threshold * _scaleFactor);
+
+			//scale factor is ignored for constant variables!
+			BDDExtensions::ShouldBeEqualTo(_y3->GetComparisonThreshold(), threshold);
```
Comment (msevestre) [inline tests/OSPSuite.SimModel.Tests/src/SimulationSpecs.cpp]: What is a constant variable?

Comment (Yuri05) [inline tests/OSPSuite.SimModel.Tests/src/SimulationSpecs.cpp]: Constant variable does not change along time and is eliminated from the ODE system
(it's just filled with is initial value).
E.g. all non-floating enzymes, transporters etc. will be const.

Actually AbsTol does not matter for such variables at all: All values are equal to the initial value of the variable defined in PK-Sim or MoBi

File: src/OSPSuite.SimModel/src/Species.cpp (hunk 1 above)
Comment (Yuri05) [inline src/OSPSuite.SimModel/src/Species.cpp]: >Are we sure that SetComparisonThreshold is called BEFORE RescaleValues?

Per construction. This is called at the end of ODE solving routine and before the values are rescaled

**Comment by Yuri05** [comment]
>I would assume that threshold should not depend on internal scale factor. We are comparing unscaled values

Threshold (based on AbsTol) is defined for the **_values ODE solver operates with_** (those scaled with the scaled factor).

E.g. imagine the following:
* ODE variable which oscillates around **1e-10**. 
* we define scale factor of **1e-10**, then the solver will operate with the values around `1e-10/1e-10 = 1`
* Absolute Tolerance is e.g. **1e-3**

For the **_scaled_** variable the threshold would be `10*AbsTol = 1e-2`
If we don't rescale the threshold, than the **_unscaled_** variable will be completely below the threshold

**Comment by msevestre** [comment]
Ok we are talking about the same thing here. I completely agree. That means
that the input threshold is scaled twice (one up and one down)?

I don't understand WHY we scale the threshold at the first place. Is it
used in the simulation? We are going to use the threshold to compare OUTPUT
values in their original scale
In your example, the values in the csv/json will be around 1e-10. So using
a scaled threshold of 1e-2 is pointless.

**Comment by Yuri05** [comment]
>In your example, the values in the csv/json will be around 1e-10. So using a scaled threshold of 1e-2 is pointless.

Exactly, using the threshold **1e-2** is pointless. But **1e-2** is the **_unscaled_** threshold. The **_scaled_** will be `1e-2 * 1e-10 `

**Comment by msevestre** [comment]
Well I do not see were the threshold is being set to 1e-2 then. There is
only one multiplication going on, not 2

**Comment by Yuri05** [comment]
1e-2 = 10*AbsTol 
(AbsTol=1e-3)

**Comment by msevestre** [comment]
Lol  but where is the AbsTol scaled then

**Comment by Yuri05** [comment]
Nowhere :-)
In my example above it was just set to 1e-3

**Comment by Yuri05** [comment]
Hmm, after all I think, that in case we have scaling factors, the implementation is correct for ODE varaibles. But not for observer. Will check it one more time

**Comment by msevestre** [comment]
I am not sure that it is correct for ODE variable either. The factor in
that case is 10* abstol. No need for scaling

**Comment by Yuri05** [comment]
Fixes #8 Add tests for Should also add some tests for observers

**Pull request merged**